### PR TITLE
Add a kernel opening mechanism for load balancing!

### PR DESCRIPTION
**Pull request by LeeMoonCh**
Add a kernel opening mechanism for load balancing.Turn on at the node with the least number of kernels.
We can use load balancing by setting the variable c.enterprisegatewayapp.load_balance = True.
If you want to start the kernel by specifying a node, you can specify it by using KERNEL_REMOTE_HOST in the variable env.

Resolves: #1062

### CODE REVIEW

File: enterprise_gateway/services/processproxies/distributed.py [hunk 1]
```diff
@@ -27,18 +27,30 @@ class DistributedProcessProxy(RemoteProcessProxy):
...
+        env_dict = kwargs.get("env")
         await super().launch_process(kernel_cmd, **kwargs)
+        if self.load_balance:
+            self.assigned_host = self.kernel_manager.parent.kernel_on_host.min_host()
+        else:
+            self.assigned_host = self._determine_next_host()
+
+        remote_host = env_dict.get("KERNEL_REMOTE_HOST")
+        if remote_host:
+            self.assigned_host = remote_host
```
[the remark this answers is not in the document]
Comment (LeeMoonCh) [reply enterprise_gateway/services/processproxies/distributed.py:52]: Thank's! @kevin-bates 
I will solve these good proposals as soon as possible.

File: enterprise_gateway/mixins.py [hunk 2]
```diff
@@ -431,6 +433,34 @@ def kernel_headers_default(self):
     def remote_hosts_default(self):
         return os.getenv(self.remote_hosts_env, self.remote_hosts_default_value).split(",")
 
+    # load_balancing_algorithm
+    load_balancing_algorithm_env = "EG_LOAD_BALANCE_ALGORITHM"
```
Comment (kevin-bates) [inline enterprise_gateway/mixins.py:437]: Could we name this like the variable:
```suggestion
    load_balancing_algorithm_env = "EG_LOAD_BALANCING_ALGORITHM"
```

File: enterprise_gateway/mixins.py [hunk 3]
```diff
@@ -431,6 +433,34 @@ def kernel_headers_default(self):
...
+        )
+
+    @validate("load_balancing_algorithm")
+    def _validate_load_balancing_algorithm(self, proposal):
+        value = proposal["value"]
+        try:
+            assert value == "round-robin" or value == "least-connection"
+        except ValueError:
+            raise TraitError(
+                f"invalid load_balancing_algorithm value {value},not in [round-robin,least-connection]"
```
Comment (kevin-bates) [inline enterprise_gateway/mixins.py:460]: Small typos...
```suggestion
                f"Invalid load_balancing_algorithm value {value}, not in [round-robin,least-connection]"
```

File: enterprise_gateway/services/processproxies/distributed.py [hunk 4]
```diff
@@ -122,12 +160,26 @@ def _build_startup_command(self, argv_cmd, **kwargs):
...
+    def _determine_next_host(self, env_dict):
+        """Simple round-robin index into list of hosts or use least-connection ."""
+        remote_host = env_dict.get("KERNEL_REMOTE_HOST")
+        if self.least_connection:
+            next_host = DistributedProcessProxy.kernel_on_host.min_or_remote_host(remote_host)
+            DistributedProcessProxy.kernel_on_host.add_kernel_id(next_host, self.kernel_id)
+        else:
+            next_host = (
+                remote_host
+                if remote_host
```
Comment (kevin-bates) [inline enterprise_gateway/services/processproxies/distributed.py:172]: Nice - thank you.

File: enterprise_gateway/mixins.py [hunk 5]
```diff
@@ -431,6 +433,34 @@ def kernel_headers_default(self):
...
         return os.getenv(self.remote_hosts_env, self.remote_hosts_default_value).split(",")
 
+    # load_balancing_algorithm
+    load_balancing_algorithm_env = "EG_LOAD_BALANCE_ALGORITHM"
+    load_balancing_algorithm_default_value = "round-robin"
+    load_balancing_algorithm = Unicode(
+        load_balancing_algorithm_default_value,
+        config=True,
+        help="""Which load balancing algorithm does DistributedProcessProxy used.
+            Default round-robin,another use least-connection.
```
Comment (kevin-bates) [inline enterprise_gateway/mixins.py:443]: Just some minor rephrasing...
```suggestion
        help="""Specifies which load balancing algorithm DistributedProcessProxy should use.
            Must be one of "round-robin" or "least-connection".
```
The default of "round-robin" will be specified in the output when running `jupyter enterprisegateway --help-all`

**Review by kevin-bates — COMMENTED** [review]
Excellent - just a couple of minor nits.

Do you want to take a shot at the documentation?  I'm thinking because this is introducing another load-balancing algorithm and given we don't currently talk about there being multiple choices, it might be easier if I added the docs in another PR since there may be "section organization" involved.  However, if you'd like to do that as part of this PR, I'd be happy to work with you.

**Comment by LeeMoonCh** [comment]
Thank you! @kevin-bates .
My English is not very good, so I won't participate in the maintenance of the document, so as to avoid a bunch of grammar problems in the document. 
5bff4784-9fbb-4e16-8c16-8be953cc95aa

**Review by kevin-bates — APPROVED** [review]
Thank you @LeeMoonCh!
I will follow up with a docs PR next week.

**Pull request merged**